src/haskell/duck.hs

```diff
@@ -0,0 +1,41 @@
+{- A typeclass that Wolf and Mallard will have to explicitly instantiate -}
+class Duck a where
+    quake :: a -> String
+    dance :: a -> String
+
+data Wolf = Wolf
+
+eat :: (Duck t) => Wolf -> t -> String
+eat _ _ = " 😈 "
+
+{- In order to become a duck, Wolf has to use the Adapter pattern. In
```

Comment:
After checking a bit about coercion in Haskell, I don't think it can be applied to this use case (or at least, not in a straight-forward way). Duck would have to be a type (impossible to coerce from a type (Wolf) to a type constraint (Duck)). And even then, it would mean making usual functions with pattern matching, and we would lose the information we have about Wolf being a Wolf.
A (untested) workaround would be to make Wolf, Duck and Mallard records, but I don't know if coercion would still be available, and if it would solve the problem adequatly. Plus, this would be ugly.

TL; DR : I don't think coercion is applicable in this case. (Though I'm far from mastering every aspect of Haskell type system).